kernel/xpmem_attach.c

```diff
@@ -162,6 +162,120 @@ xpmem_close_handler(struct vm_area_struct *vma)
 #endif
 }
 
+/* Find the range of non present entries */
+static u64
+xpmem_fault_vaddr_end(u64 vm_end, u64 addr, unsigned long max_pages)
+{
+	for (; max_pages; max_pages--, addr += PAGE_SIZE) {
+		if (addr >= vm_end) {
+			break;
+		}
+		if (xpmem_vaddr_to_PFN(current->mm, addr)) {
+			break;
+		}
+	}
+	return addr;
+}
+
+/* Find the range of non present entries before the request address */
+static u64
+xpmem_fault_vaddr_start(u64 vm_start, u64 addr, unsigned long max_pages)
+{
+	for (; max_pages; max_pages--, addr -= PAGE_SIZE) {
+		if (addr <= vm_start) {
+			break;
+		}
+		if (xpmem_vaddr_to_PFN(current->mm, addr - PAGE_SIZE)) {
+			break;
+		}
+	}
+	return addr;
+}
+
+static int
+xpmem_fault_pages(struct xpmem_segment *seg, struct vm_area_struct *vma,
+		  u64 vaddr)
+{
+	int ret;
+	u64 seg_vaddr;
+	u64 end;
+	u64 start;
+	struct xpmem_attachment *att;
+	unsigned long nr_pages;
+	struct vm_area_struct *src_vma;
+	unsigned long pages_after;
+	unsigned long pages_before;
+	struct page *pages[XPMEM_MAX_PAGE_FAULTS];
+	int result = 0;
+	xpmem_max_page_fault_get(&pages_before, &pages_after);
+
+	att = vma->vm_private_data;
+	if (!att) {
+		goto out;
+	}
+
+	if (vaddr < att->at_vaddr || vaddr + 1 > att->at_vaddr + att->at_size) {
+		goto out;
+	}
+	if (vaddr < vma->vm_start || vaddr >= vma->vm_end) {
+		goto out;
+	}
+
+	end = min((u64)vma->vm_end, att->at_vaddr + att->at_size);
+	start = max((u64)vma->vm_start, att->at_vaddr);
+
+	/*
+	 * When calling memcpy(), glibc can use memmove() instead, which can fault
+	 * pages backward, so we have following possible faulting patterns:
+	 * - Random: start <= vaddr and end > vaddr
+	 * - Forward: start == vaddr as previous PTE's are present, end > vaddr
+	 * - Backward: end == vaddr + 1, as PTE's after are present, start <= vaddr
+	 */
+	end = xpmem_fault_vaddr_end(end, vaddr + PAGE_SIZE, pages_after);
+	start = xpmem_fault_vaddr_start(start, vaddr, pages_before);
+
+	XPMEM_DEBUG("vaddr = %llx, start = %llx, end = %llx", vaddr, start,
+		    end);
+
+	while (start < end) {
+		/* translate current fault vaddr to the source vaddr */
+		seg_vaddr = (att->vaddr & PAGE_MASK) + (start - att->at_vaddr);
+
+		nr_pages = (end - start) >> PAGE_SHIFT;
+		BUG_ON(nr_pages > XPMEM_MAX_PAGE_FAULTS);
+
+		ret = xpmem_ensure_valid_PFN(seg, seg_vaddr, pages, nr_pages);
+		if (ret >= 0) {
+			att->flags |= XPMEM_FLAG_VALIDPTEs;
+		}
+		if (ret <= 0) {
```

Comment:
why would xpmem_ensure_valid_PFN fail, given that we calculated start,end based of ranges of valid PTEs?

Reply:
there is:
- if the vma is an xpmem (vm_ops == xpmem_vm_ops)
- get_user_pages() can return < 0
- and if segment has been marked as destroying